src/ipips/ipip-0484.md

```diff
@@ -0,0 +1,102 @@
+
+The Delegated Routing v1 API empowers resource constrained clients like web browsers by significantly reducing the number of network connections necessary to fetch content addressed data directly from provider peers.
+
+However, there are many cases where most of the results from the Delegated Routing v1 API are not actionable by clients, because the client does not support either the **network transport protocol** or the **transfer protocol** of the provider.
+
+For instance, web browsers are limited to a specific set of network transport protocols, namely HTTPS, Secure WebSockets, WebTransport (emerging), and WebRTC. Consequently, providing information about providers that exclusively support TCP and/or UDP is not beneficial for browser-based clients, as they are unable to utilize such connections.
+
+Moreover, [Helia](https://github.com/ipfs/helia/), the most actively maintained browser IPFS implementation, supports block retrieval by CID with Bitswap and Trustless Gateways, but does not support GraphSync.
+
+This means that returning providers that only support raw TCP, raw UDP/QUIC, or GraphSync from the Delegated Routing API is not useful for browser clients, and results in unnecessary network traffic for browser clients.
+
+## Note on terminology
+
+The term **"transport"** is overloaded in the IPFS ecosystem.
+
+In the context of this IPIP, we refer to the network layer transport protocol, e.g. TCP, QUIC, WebTransport, as **"network transport protocol"** to avoid ambiguity.
+
+**"Transfer protocol"** refers to data transfer protocols, i.e. content-addressed block retrieval protocols, e.g. Bitswap, GraphSync, HTTP.
+
+## Detailed design
+
+### Network Transport Protocol Filtering
+
+The proposed change is to add a `?filter-addrs` parameter to the `GET /routing/v1/providers/{cid}` endpoint of :cite[http-routing-v1]:
+
+- Add a `?filter-addrs=<comma-separated-list>` optional parameter to `GET /routing/v1/providers/{CID}` that indicates which network transports to return by filtering the multiaddrs in the `Addrs` field of the [Peer schema].
+- The value of the `filter-addrs` parameter is a comma-separated list of network transport protocol _name strings_ as defined in the [multiaddr protocol registry](https://github.com/multiformats/multiaddr/blob/master/protocols.csv), e.g. `?filter-addrs=webtransport`.
+- `unknown` can be be passed to include providers whose multiaddrs are unknown, e.g. `?filter-addrs=unknown`. This allows filtering providers whose multiaddrs are unknown at the time of filtering.
+- Multiaddrs are filtered by checking if the protocol name appears in any of the multiaddrs (logical OR).
+- Only multiaddrs that pass the filter are included in the response.
+- If there are no multiaddrs that match the passed transports, the provider is omitted from the response.
+- Negative filtering is done by prefixing the protocol name with `!`, e.g. `?filter-addrs=!quic-v1,!tcp`.
```

Comment:
negative filtering with logical OR will cause problems. 

For example, if we want to get all non `p2p-circuit` that are `webtransport` and `webrtc-direct` maddrs,  i.e. `?filter-addrs=!p2p-circuit,webtransport,webrtc-direct` we get something that looks like

`if (!maddr.has(p2p-circuit) || maddr.has(webtransport) || maddr.has(webrtc-direct)) { include maddr }`

which basically returns all non p2p-circuit addresses in addition to webtransport

Whereas we actually want something more like:

`if (!maddr.has(p2p-circuit) && (maddr.has(webtransport) || maddr.has(webrtc-direct)))  { include maddr }`

Reply:
would it be too complex to support grouping with `[]` or `()`?

Reply:
Can you think of examples where you would want that? Otherwise, I'd say YAGNI